packages/amd-loader/amd-loader.js

```diff
@@ -0,0 +1,196 @@
+'use strict';
+
+(function() {
+
+function Registry() {
+  this.modules = {};
+  this.topLevelIdx = 0;
+  this.previousTopLevelUrl = undefined;
+  this.pendingDefine = undefined;
+}
+
+Registry.prototype.get = function(url) {
+  var module = this.modules[url];
+  if (module === undefined) {
+    module = this.modules[url] = new Module(url);
+  }
+  return module;
+};
+
+var registry = new Registry();
+
+function getBasePath(url) {
+  // TODO(aomarks) Maybe a better way to do this.
+  var lastSlash = url.lastIndexOf('/');
+  if (lastSlash === url.indexOf('://') + 2) {
+    // http://example.com -> http://example.com/
+    return url + '/';
+  }
+  // http://example.com/foo/bar.html -> http://example.com/foo/
+  return url.substring(0, lastSlash + 1);
+}
+
+function Module(url) {
+  this.url = url;
+  this.basePath = getBasePath(url);
+  this.exports = {};
+  this.loadStartedOrDone = false;
+  // All of this module's dependencies are resolved and its factory has run.
+  this.resolved = false;
+  // Callbacks from dependents waiting to hear when this module has resolved.
+  this.notify = [];
+}
+
+var resolverAnchor = document.createElement('a');
+
+Module.prototype.init = function(deps, factory) {
+  var factoryArgs = [];
+  var numUnresolvedDeps = 0;
+
+  function onDepResolved() {
+    numUnresolvedDeps--;
+    maybeResolve();
+  }
+
+  var thisResolve = this.resolve.bind(this);
+  function maybeResolve() {
+    if (numUnresolvedDeps === 0) {
+      factory.apply(null, factoryArgs);
+      thisResolve();
+    }
+  }
+
+  for (var i = 0; i < (deps || []).length; i++) {
+    var depSpec = deps[i];
+
+    if (depSpec === 'exports') {
+      factoryArgs[i] = this.exports;
+
+    } else if (depSpec === 'require') {
+      factoryArgs[i] = this.makeDynamicImporter();
+
+    } else if (depSpec === 'meta') {
```

Comment:
AMD includes a "module" built-in dependency, but the spec doesn't seem to specify it that I can see. In IMD it has the `exports` and `id` properties. Maybe we dug that out of RequireJS source. Seems like in this loader `module.id` might === `import.meta.url`.

Reply:
Looks like the AMD spec defers to the CommonJS spec for `module`, which says:

> 1) The "module" object must have a "id" property that is the top-level "id" of the module. The "id" property must be such that require(module.id) will return the exports object from which the module.id originated. (That is to say module.id can be passed to another module, and requiring that must return the original module). When feasible this property should be read-only, don't delete.
> 
> 2) The "module" object may have a "uri" String that is the fully-qualified URI to the resource from which the module was created. The "uri" property must not exist in a sandbox.
> 
> http://wiki.commonjs.org/wiki/Modules/1.1.1#Module_Context

I'm open to using `module.id` or `module.uri` (or both). Although, I do also like the idea of providing it via `meta`, because we are trying to match the semantics of `import.meta`. I don't know if `import.meta.scriptElement` is going to happen, but if it did, then we would have a natural place to put it. I think this discussion depends on what our position is on interopability with other AMD modules/loaders.